### DISCUSSION TITLE
nixos-with-flakes/nixos-flake-and-module-system

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#157] [Announcements]
# nixos-with-flakes/nixos-flake-and-module-system

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixos-with-flakes/nixos-flake-and-module-system

**Comment by RobotDisco** [comment]
Finally! Someone does a concise explanation of `specialArgs` and `_module.args`, which I have only see as mystery magic that I've needed to reverse engineer in fear.

**Comment by BlastboomStrice** [comment]
Hey hello and thanks for this very detailed guide. As I'm re-reading it after about a year, I noticed that in first steps of setting up the `flake.nix` you use this line:

```nix
outputs = { self, nixpkgs, ... }@inputs: {
```

while later (on this page) you switch to this line:

```nix
outputs = inputs@{ self, nixpkgs, ... }: {
```

Is there any difference? Could you maybe add some explanation in the code?

**Reply by ryan4yin** [reply]
These two syntax forms are semantically equivalent — it's a fundamental concept you'll encounter while learning the Nix language. I'd highly recommend checking out the official tutorial at https://nix.dev/tutorials/nix-language.

**Comment by ask-elad** [comment]
damm dudee you are savior!!
thanks for your valuable contribution

**Comment by jliklik** [comment]
Wow, thanks for the amazing guide Ryan!
